rust/stub/src/pe_loader.rs

```diff
@@ -0,0 +1,91 @@
+use goblin::pe::PE;
+use uefi::{
+    prelude::BootServices,
+    table::{
+        boot::{AllocateType, MemoryType},
+        Boot, SystemTable,
+    },
+    Handle, Status,
+};
+
+pub struct Image {
+    pub base: *const u8,
+    pub length: usize,
+    pub entry: extern "efiapi" fn(Handle, SystemTable<Boot>) -> Status,
+}
+
+#[cfg(target_arch = "x86_64")]
+fn flush_instruction_cache(_start: *const u8, _length: usize) {
+    // x86_64 mandates coherent instruction cache
+}
+
+/// Loads and relocates a trusted PE file, returning a structure containing its allocated virtual memory range and entry point.
+pub unsafe fn load(boot_services: &BootServices, file_data: &[u8]) -> uefi::Result<Image> {
+    let pe = PE::parse(file_data).expect("Malformed PE file");
+
+    // Allocate all memory the image will need in virtual memory.
+    // We follow shim here and allocate as EfiLoaderCode.
+    let mut pos = 0;
+    for section in &pe.sections {
+        assert!(section.virtual_address >= pos);
+        pos = section.virtual_address + section.virtual_size;
+    }
+    let base = boot_services.allocate_pages(
```

Comment:
`Image` should free the allocated memory on `Drop`. Then it is also safe to use a `&[u8]` slice instead of pointer + size?

Reply:
Or even better, have a small RAII wrapper around `allocate_pages` so we don't leak memory when we fail to load the image.

Reply:
I don't see this working since deallocating pages requires boot services. However, I implemented deallocating the kernel if it returns.

Reply:
Yes, that's what I meant!